ext/opcache/shared_alloc_mmap.c

```diff
@@ -40,6 +40,340 @@
 # define MAP_HUGETLB MAP_ALIGNED_SUPER
 #endif
 
+#if defined(__x86_64__) && defined(__linux__)
+#define max(a,b) ((a) > (b) ? (a):(b))
+
+/* This function must exist in PHP .text segment */
+void* php_text_lighthouse(void);
+
+#if defined(MAP_HUGETLB)
+static const size_t huge_page_size = 2 * 1024 * 1024; /* 2MB page size */
+#else
+static const size_t ordinary_page_size = 4 * 1024;    /* 4KB page size */
+#endif /* MAP_HUGETLB */
+
+static const ptrdiff_t addr_4GB_distance = UINT32_MAX; /* 4GB address distance */
+static const uintptr_t addr_4GB = UINT32_MAX;          /* 4GB memory address */
+static const size_t    size_4GB = UINT32_MAX;          /* 4GB memory size */
+
+/* Store the segment address information */
+struct map_seg_addresses {
+    uintptr_t start; /* segment start address (included) */
+    uintptr_t end;   /* segment end address (excluded) */
+};
+
+/* Count segments in /proc/self/maps file; each line represents one segment. */
+static size_t map_segment_count(void)
```

Comment:
If I understand correctly, you are counting and parsing in two separate steps to avoid allocations during parsing, which could change the map ? This might be worth a comment :)

Reply:
Yes we want to use array rather than malloc. Will add some more comments.

Reply:
Using malloc() might lead to change of maps file. will add comment in new patch.